internal/status.go

```diff
@@ -0,0 +1,147 @@
+package internal
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/dustin/go-humanize"
+)
+
+type StatusLine struct {
+	resources []Resource
+
+	resourceSizes       []int64
+	totalBytes          int64
+	sizingSuccess       bool
+	bytesDownloaded     int64
+	resourcesDownloaded int
+	spinI               int
+	indexCh             chan int
+	startTime           time.Time
+}
+
+var spinChars = [5]string{"-", "\\", "|", "/", "-"}
+
+// Create, initialize, and start a new StatusLine.
+func NewStatusLine(resources []Resource) *StatusLine {
+	st := StatusLine{}
+	st.resources = resources
+	st.indexCh = make(chan int)
+	st.resourceSizes, st.totalBytes, st.sizingSuccess = getResourcesSizes(resources, 10000)
+
+	st.start()
+
+	return &st
+}
+
+// Inform StatusLine that resource has finished downloading.
+func (st *StatusLine) increment(i int) {
+	st.indexCh <- i
+}
+
+// Begin goroutine and for loop.
+func (st *StatusLine) start() {
+	st.startTime = time.Now()
+
+	go func() {
+		st.spinI = 0
+
+		for {
+
+			//Block until value is inserted into indexCh (>= 0 when resource done downloading, -1 every 50 milliseconds to keep timer and spinner updating).
+			var i int
+			select {
+			case i = <-st.indexCh:
+			case <-time.After(50 * time.Millisecond):
+				i = -1
+			}
+			if i != -1 {
+				st.bytesDownloaded += st.resourceSizes[i]
+				st.resourcesDownloaded++
+			}
+
+			//Update spinner.
+			st.spinI += 1
+			if st.spinI == len(spinChars) {
+				st.spinI = 0
+			}
+
+			// Print line.
+			anyRemaining := st.resourcesDownloaded < len(st.resources)
+			line := composeStatusString(st.bytesDownloaded, st.totalBytes, st.resourcesDownloaded, len(st.resources), st.sizingSuccess, spinChars[:], st.spinI, st.startTime, anyRemaining, 20)
+			fmt.Print(line)
+
+			//Exit if done.
+			if !anyRemaining {
+				fmt.Println()
+				break
+			}
+
+		}
+	}()
+
+}
+
+func getResourcesSizes(resources []Resource, timeoutMs int) ([]int64, int64, bool) {
```

Comment:
I suggest refactoring this as a method on StatusLine with the following signature: `func (st *StatusLine) initSizes() error`.
Making it a method you avoid having to pass around things that you end up just setting on the object itself anyway.
I think the timeout should be a module level const, no need to pass it around as a parameter.
It's more idiomatic to return an `error` (see examples in the code) rather than a bool.